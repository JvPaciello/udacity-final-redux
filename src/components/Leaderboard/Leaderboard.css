body, html {
  margin: 0;
  padding: 0;
  width: 100%;
  height: 100%;
  overflow-x: hidden; 
  background-color: #000000;
  font-family: Arial, sans-serif;
  box-sizing: border-box;
}

.leaderboard {
  max-width: 800px;
  margin: 5rem auto;
  padding: 20px;
  text-align: center;
  background-color: #ffffff;
  border-radius: 20px;
  box-shadow: 0 4px 8px rgba(0, 0, 0, 0.2);
}

.leaderboard h1 {
  font-size: 28px;
  margin-bottom: 30px;
  color: #14171a;
  font-weight: bold;
}

table {
  width: 100%;
  border-collapse: collapse;
  margin: 0 auto;
  background-color: #f9f9f9;
  border-radius: 20px;
  overflow: hidden;
  box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
}

th, td {
  padding: 15px;
  text-align: center;
}

th {
  background-color: #000000;
  color: white;
  font-size: 18px;
  font-weight: bold;
}

tr {
  transition: background-color 0.3s ease, transform 0.3s ease;
}

tr:nth-child(even) {
  background-color: #f2f2f2;
}

tr:hover {
  background-color: #cacbcc;
  color: white;
  transform: scale(1.02);
}

.avatar {
  width: 85px;
  height: 85px;
  border-radius: 50%;
  object-fit: cover;
}

td {
  font-size: 16px;
  color: #333;
  font-weight: 500;
  border-bottom: 1px solid #ddd;
}

td:last-child {
  font-weight: bold;
  color: #000000;
}

/* Responsividade adicional */
@media (max-width: 768px) {
  .leaderboard {
    padding: 10px;
    margin: 2rem auto;
  }

  table {
    font-size: 14px;
    border: 0; /* Remove bordas para telas menores */
  }

  thead {
    display: none; /* Oculta cabeçalho da tabela no mobile */
  }

  tr {
    display: block;
    margin-bottom: 15px;
    box-shadow: 0 2px 5px rgba(0, 0, 0, 0.1); /* Estilo de card */
  }

  td {
    display: flex;
    justify-content: space-between;
    padding: 10px;
    font-size: 14px;
    border-bottom: 1px solid #ddd;
  }

  td img.avatar {
  display: block; 
  margin: 0 auto; 
  }

  td::before {
    content: attr(data-label); /* Adiciona rótulos nas células */
    font-weight: bold;
    color: #000;
    margin-right: 10px;
  }

  .avatar {
    width: 60px;
    height: 60px;
  }
}

