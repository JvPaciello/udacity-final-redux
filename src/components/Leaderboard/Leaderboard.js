import React from "react";
import { useSelector } from "react-redux";
import "./Leaderboard.css";

function Leaderboard() {
  const users = useSelector((state) => state.users);

  // Calculate scores for each user
  const leaderboardData = Object.values(users)
    .map((user) => ({
      id: user.id,
      name: user.name,
      avatar: user.avatarURL,
      questionsAsked: user.questions.length,
      questionsAnswered: Object.keys(user.answers).length,
      totalScore: user.questions.length + Object.keys(user.answers).length,
    }))
    .sort((a, b) => b.totalScore - a.totalScore); // Sort by total score descending

  return (
    <div className="leaderboard">
      <h1>Leaderboard</h1>
      <table>
        <thead>
          <tr>
            <th>Avatar</th>
            <th>Name</th>
            <th>Answered</th>
            <th>Created</th>
            <th>Score</th>
          </tr>
        </thead>
        <tbody>
          {leaderboardData.map((user) => (
            <tr key={user.id}>
              <td>
                <img
                  src={user.avatar}
                  alt={`Avatar of ${user.name}`}
                  className="avatar"
                />
              </td>
              <td>{user.name}</td>
              <td>{user.questionsAnswered}</td>
              <td>{user.questionsAsked}</td>
              <td>{user.totalScore}</td>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
}

export default Leaderboard;
