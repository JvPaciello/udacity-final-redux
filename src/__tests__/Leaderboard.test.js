/* eslint-disable testing-library/prefer-screen-queries */
import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import Leaderboard from "../components/Leaderboard/Leaderboard";

const mockStore = configureStore([]);
const initialState = {
  users: {
    sarahedo: {
      id: "sarahedo",
      name: "Sarah Edo",
      avatarURL:
        "https://austinmonthly.wppcdn.com/wp-content/uploads/2019/10/JackieVenson1.jpeg",
      answers: {
        "8xf0y6ziyjabvozdd253nd": "optionOne",
        "6ni6ok3ym7mf1p33lnez": "optionOne",
        "am8ehyc8byjqgar0jgpub9": "optionTwo",
        "loxhs1bqm25b708cmbf3g": "optionTwo",
      },
      questions: ["8xf0y6ziyjabvozdd253nd", "am8ehyc8byjqgar0jgpub9"],
    },
    tylermcginnis: {
      id: "tylermcginnis",
      name: "Tyler McGinnis",
      avatarURL:
        "https://blogs.correiobraziliense.com.br/trilhasonora/wp-content/uploads/sites/39/2020/10/CBNFOT081020100047-550x549.jpg",
      answers: {
        "vthrdm985a262al8qx3do": "optionOne",
        "xj352vofupe1dqz9emx13r": "optionTwo",
      },
      questions: ["loxhs1bqm25b708cmbf3g", "vthrdm985a262al8qx3do"],
    },
  },
};

describe("Leaderboard Component", () => {
  it("matches the snapshot", () => {
    const store = mockStore(initialState);
    const { asFragment } = render(
      <Provider store={store}>
        <Leaderboard />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("orders users correctly by totalScore in descending order", () => {
    const store = mockStore({
      users: {
        user1: {
          id: "user1",
          name: "User One",
          avatarURL: "avatar1.png",
          answers: { question1: "optionOne" },
          questions: ["question2"],
        },
        user2: {
          id: "user2",
          name: "User Two",
          avatarURL: "avatar2.png",
          answers: { question1: "optionOne", question2: "optionTwo" },
          questions: ["question3", "question4"],
        },
      },
    });

    const { getAllByText } = render(
      <Provider store={store}>
        <Leaderboard />
      </Provider>
    );

    const names = getAllByText(/User (One|Two)/).map((node) => node.textContent);
    expect(names).toEqual(["User Two", "User One"]); 
  });

  it("renders avatars for all users", () => {
    const store = mockStore({
      users: {
        user1: {
          id: "user1",
          name: "User One",
          avatarURL: "avatar1.png",
          answers: {},
          questions: [],
        },
        user2: {
          id: "user2",
          name: "User Two",
          avatarURL: "avatar2.png",
          answers: {},
          questions: [],
        },
      },
    });

    const { getByAltText } = render(
      <Provider store={store}>
        <Leaderboard />
      </Provider>
    );

    expect(getByAltText("Avatar of User One").src).toContain("avatar1.png");
    expect(getByAltText("Avatar of User Two").src).toContain("avatar2.png");
  });
  it("renders the correct table headers", () => {
    const store = mockStore(initialState);
  
    const { getByText } = render(
      <Provider store={store}>
        <Leaderboard />
      </Provider>
    );
  
    expect(getByText("Avatar")).toBeInTheDocument();
    expect(getByText("Name")).toBeInTheDocument();
    expect(getByText("Answered")).toBeInTheDocument();
    expect(getByText("Created")).toBeInTheDocument();
    expect(getByText("Score")).toBeInTheDocument();
  });
  
  it("renders an empty state when no users are present", () => {
    const store = mockStore({ users: {} });
  
    const { getByText } = render(
      <Provider store={store}>
        <Leaderboard />
      </Provider>
    );
  
    expect(getByText("Leaderboard")).toBeInTheDocument();
    expect(getByText("No data available")).toBeInTheDocument();
  });
  
});
