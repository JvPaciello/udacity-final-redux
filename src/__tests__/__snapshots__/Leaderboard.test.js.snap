// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leaderboard Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="leaderboard"
  >
    <h1>
      Leaderboard
    </h1>
    <table>
      <thead>
        <tr>
          <th>
            Avatar
          </th>
          <th>
            Name
          </th>
          <th>
            Answered
          </th>
          <th>
            Created
          </th>
          <th>
            Score
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <img
              alt="Avatar of Sarah Edo"
              class="avatar"
              src="https://austinmonthly.wppcdn.com/wp-content/uploads/2019/10/JackieVenson1.jpeg"
            />
          </td>
          <td>
            Sarah Edo
          </td>
          <td>
            4
          </td>
          <td>
            2
          </td>
          <td>
            6
          </td>
        </tr>
        <tr>
          <td>
            <img
              alt="Avatar of Tyler McGinnis"
              class="avatar"
              src="https://blogs.correiobraziliense.com.br/trilhasonora/wp-content/uploads/sites/39/2020/10/CBNFOT081020100047-550x549.jpg"
            />
          </td>
          <td>
            Tyler McGinnis
          </td>
          <td>
            2
          </td>
          <td>
            2
          </td>
          <td>
            4
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;

exports[`Leaderboard Snapshot Test renders correctly with given state 1`] = `
<div
  className="leaderboard"
>
  <h1>
    Leaderboard
  </h1>
  <table>
    <thead>
      <tr>
        <th>
          Avatar
        </th>
        <th>
          Name
        </th>
        <th>
          Answered
        </th>
        <th>
          Created
        </th>
        <th>
          Score
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          <img
            alt="Avatar of Sarah Edo"
            className="avatar"
            src="https://austinmonthly.wppcdn.com/wp-content/uploads/2019/10/JackieVenson1.jpeg"
          />
        </td>
        <td>
          Sarah Edo
        </td>
        <td>
          4
        </td>
        <td>
          2
        </td>
        <td>
          6
        </td>
      </tr>
      <tr>
        <td>
          <img
            alt="Avatar of Tyler McGinnis"
            className="avatar"
            src="https://blogs.correiobraziliense.com.br/trilhasonora/wp-content/uploads/sites/39/2020/10/CBNFOT081020100047-550x549.jpg"
          />
        </td>
        <td>
          Tyler McGinnis
        </td>
        <td>
          2
        </td>
        <td>
          2
        </td>
        <td>
          4
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Leaderboard Tests renders correctly with given state 1`] = `
<div
  className="leaderboard"
>
  <h1>
    Leaderboard
  </h1>
  <table>
    <thead>
      <tr>
        <th>
          Avatar
        </th>
        <th>
          Name
        </th>
        <th>
          Answered
        </th>
        <th>
          Created
        </th>
        <th>
          Score
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          <img
            alt="Avatar of Sarah Edo"
            className="avatar"
            src="https://austinmonthly.wppcdn.com/wp-content/uploads/2019/10/JackieVenson1.jpeg"
          />
        </td>
        <td>
          Sarah Edo
        </td>
        <td>
          4
        </td>
        <td>
          2
        </td>
        <td>
          6
        </td>
      </tr>
      <tr>
        <td>
          <img
            alt="Avatar of Tyler McGinnis"
            className="avatar"
            src="https://blogs.correiobraziliense.com.br/trilhasonora/wp-content/uploads/sites/39/2020/10/CBNFOT081020100047-550x549.jpg"
          />
        </td>
        <td>
          Tyler McGinnis
        </td>
        <td>
          2
        </td>
        <td>
          2
        </td>
        <td>
          4
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
